### ISSUE TITLE
Roadmap to 5.0

### ISSUE THREAD

**Issue by jordisala1991** [sonata-project/SonataUserBundle#1465]
## Feature Request

There are more things to do for 5.0:

- [x] Avoid using parameters on container as much as possible
- [x] User form types instead of hardcoded forms
- [x] Html support in email
- [x] Fix UserAdmin only change password (add tests similar to MediaBundle)
- [x] Add commands (adapt commands from FOSUserBundle)
- [x] Psalm level 2
- [x] Deprecate/Remove Groups feature
- [x] Update documentation
- [x] Improve configuration #1464 https://github.com/sonata-project/SonataUserBundle/pull/1481
- [x] Mark final classes
- [x] Move from public/protected to private
- [x] Avoid direct usage of container https://github.com/sonata-project/SonataUserBundle/pull/1481
- [x] Simplify impersonating routing configuration https://github.com/sonata-project/SonataUserBundle/pull/1481
- [x] Check twig templates
- [x] Replace glyphicons by font awesome icons
- [x] Improve tests (mark final, remove legacy)
- [x] Use array instead of json for roles

Translations:
- [x] Add missing translations for english
- [x] Add missing translations for rest of languages (FOSUser)
- [x] Remove untranslated translations
- [x] Add missing validation translations (FOSUser)

Tests:
- [x] Test migration from SonataUser v4 to v5 (previous user data works) (4.4)
- [x] Test all features from the bundle on new security system (6.0)
- [x] Test all features from the bundle on the old security system (5.4)

Documentation:
- [x] Improve documentation for Symfony 5/6 installs

As soon as I know more things to do I will keep adding them.

**Comment by andyhobbs** [comment]
Will be Groups feature in another bundle?

**Comment by jordisala1991** [comment]
If you just need to group some roles you can use roles_hierarchy from symfony already.

If you need some more complex solution, you might take a look at voters. The groups feature was not a good solution since it forces always a query to check for the groups roles, it wont be provided by Sonta (you can still reimplement it by yourself if you really need to)

**Comment by jordisala1991** [comment]
After #1490 we are almost there

Things missing:
* Fixing translations (we have to migrate some translations from FOSUser on other languages, and check if there are placeholder translations and things like that)
* Improving documentation for Symfony 5 / 6 (security system is different so we might provide configuration with both Symfony versions)

Even though we have a lot of functional tests and I'm pretty sure most of the bundle should work just fine if would be nice to also test this with a real app on Symfony (more than one version ideally 5.4 + 6.0 + new security system / old security system)

There is one nice to have, which is to be able to disable/enable reseting password feature. Tbh I don't care a lot if we don't do it for 5.0, it wasn't configurable before anyway.

The only thing that bothers me a bit is: #1483 . If that's true we might take a look at that, because otherwise the migration won't be easy at all, you might need to change all passwords coming from FOSUser to make your previous data work. It would be nice to have this working but not sure if it is possible.

wdyt @VincentLanglet ?

We might do an alpha release, but with #1483 I am not sure about it.

**Comment by VincentLanglet** [comment]
> There is one nice to have, which is to be able to disable/enable reseting password feature. Tbh I don't care a lot if we don't do it for 5.0, it wasn't configurable before anyway.
>
> The only thing that bothers me a bit is: https://github.com/sonata-project/SonataUserBundle/issues/1483 . If that's true we might take a look at that, because otherwise the migration won't be easy at all, you might need to change all passwords coming from FOSUser to make your previous data work. It would be nice to have this working but not sure if it is possible.

I totally agree with you

> We might do an alpha release, but with https://github.com/sonata-project/SonataUserBundle/issues/1483 I am not sure about it.

The alpha could be a way to have more early-tester and confirm the issue or get help on it.

**Comment by jordisala1991** [comment]
> > There is one nice to have, which is to be able to disable/enable reseting password feature. Tbh I don't care a lot if we don't do it for 5.0, it wasn't configurable before anyway.
> > The only thing that bothers me a bit is: #1483 . If that's true we might take a look at that, because otherwise the migration won't be easy at all, you might need to change all passwords coming from FOSUser to make your previous data work. It would be nice to have this working but not sure if it is possible.
> 
> I totally agree with you
> 
> > We might do an alpha release, but with #1483 I am not sure about it.
> 
> The alpha could be a way to have more early-tester and confirm the issue or get help on it.

I have extracted the reset password optional feature and merged the last PR.

To me we are ready for an alpha to see if there is a real problem with those passwords.

**Comment by VincentLanglet** [comment]
We just need to generate the changelog then

**Comment by jordisala1991** [comment]
Besides the problem solved in #1500 it works good on a Symfony 6 with new security system so far. I have tested the password migration from RunroomUserBundle but didn't had the time to test it coming from FOSUserBundle (I might need some help with that)

**Comment by jordisala1991** [comment]
I have tested for the old security system, it works too. You need to make sure you disable:

```
security:
    enable_authenticator_manager: false
```

And enabled anonymous:

```
security:
    firewalls:
        admin:
            anonymous: true
```

**Comment by VincentLanglet** [comment]
So we just need to update the doc before the next release ?

**Comment by jordisala1991** [comment]
There are 3 things missing:

- Add missing translations for rest of languages (FOSUser)
- Test migration from SonataUser v4 to v5 (previous user data works) (4.4)
-  Improve documentation for Symfony 5/6 installs

**Comment by 4c0n** [comment]
I might be able to test the migration from v4 to v5 soon, hopefully this week or the next.

**Comment by jordisala1991** [comment]
I did a test for the migration from SF 4.4 Sonata 3 to SF 4.4 Sonata 4 (including migration from SonataUserBundle 4 to 5).
My steps was:

* Clean composer.json with only flex
* Install framework-bundle, sonata-user, sonata-admin and sonata-doctrine-orm
* Configure all bundles with the security part, entities and all, using bcrypt for the password
* Create the database and a user with `console fos:user:create --super-admin`
* Ensure I can log in

Then for the update:

* Modify composer.json to install Sonata v4 and SonataUserBundle v5
* Remove FOSUserBundle related config
* Update security configuration (this will need a good documentation update, some parts of the old docs are not even right)
* Update the schema (multiple fields on the user table are removed)
* Logout and try to login again

Everything worked so far... if someone else want to try to see if there is a bug.. but for my side I can't reproduce with those steps.

**Comment by jordisala1991** [comment]
Since we are only missing translations and docs, we might do an RC release? wdyt @VincentLanglet , or should we wait for translations at least?

**Comment by VincentLanglet** [comment]
> Since we are only missing translations and docs, we might do an RC release? wdyt @VincentLanglet , or should we wait for translations at least?

I would say that having translations could be better for a RC
